contracts/service-registry/src/contract.rs

```diff
@@ -163,126 +264,152 @@ pub mod execute {
         Ok(Response::new())
     }
 
-    pub fn deregister_worker(
+    pub fn declare_chain_support(
+        deps: DepsMut,
+        info: MessageInfo,
+        service_name: String,
+        chains: Vec<String>,
+    ) -> Result<Response, ContractError> {
+        SERVICES
+            .may_load(deps.storage, &service_name)?
+            .ok_or(ContractError::ServiceNotFound {})?;
+
+        WORKERS
+            .may_load(deps.storage, (&service_name, &info.sender))?
+            .ok_or(ContractError::WorkerNotFound {})?;
+
+        for chain in chains {
+            WORKERS_PER_CHAIN.save(deps.storage, (&service_name, &chain, &info.sender), &())?;
+        }
+
+        Ok(Response::new())
+    }
+
+    pub fn unbond_worker(
         deps: DepsMut,
         env: Env,
         info: MessageInfo,
         service_name: String,
     ) -> Result<Response, ContractError> {
-        service_workers().update(
+        SERVICES
+            .may_load(deps.storage, &service_name)?
+            .ok_or(ContractError::ServiceNotFound {})?;
+
+        let worker = WORKERS
+            .may_load(deps.storage, (&service_name, &info.sender))?
+            .ok_or(ContractError::WorkerNotFound {})?;
+
+        let can_unbond = true; // TODO: actually query the service to determine this value
+
+        let bonding_state = match worker.bonding_state {
+            BondingState::Bonded { amount } | BondingState::RequestedUnbonding { amount }
+                if can_unbond =>
+            {
+                Ok(BondingState::Unbonding {
+                    unbonded_at: env.block.time,
+                    amount,
+                })
+            }
+            BondingState::Bonded { amount } if !can_unbond => {
+                Ok(BondingState::RequestedUnbonding { amount })
+            }
+            _ => Err(ContractError::InvalidBondingState(worker.bonding_state)),
+        }?;
```

Comment:
Again, this should be extracted into a method of `BondingState` `fn unbond(self) -> Self` and be unit tested.